File: cpp/src/lance/io/filter_test.cc
//  Copyright 2022 Lance Authors
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#include "lance/io/filter.h"

#include <arrow/array.h>
#include <arrow/compute/exec/expression.h>
#include <arrow/record_batch.h>
#include <fmt/format.h>

#include <catch2/catch_test_macros.hpp>

#include "lance/arrow/stl.h"
#include "lance/arrow/type.h"
#include "lance/format/schema.h"

using ::arrow::compute::equal;
using ::arrow::compute::field_ref;
using ::arrow::compute::literal;
using ::arrow::compute::or_;

const auto kSchema =
    lance::format::Schema(::arrow::schema({::arrow::field("pk", ::arrow::int32()),
                                           ::arrow::field("value", ::arrow::int32()),
                                           ::arrow::field("label", ::arrow::utf8())}));

TEST_CASE("Test without filter") {
  auto empty_filter =
      lance::io::Filter::Make(kSchema, ::arrow::compute::literal(true)).ValueOrDie();
  CHECK(empty_filter == nullptr);
}

TEST_CASE("Test with one condition") {
  auto filter = lance::io::Filter::Make(kSchema, equal(field_ref("value"), literal("32")));
  INFO("filter " << filter.status().message());
  CHECK(filter.ok());

  filter = lance::io::Filter::Make(kSchema, equal(literal("32"), field_ref("value")));
  INFO("filter " << filter.status().message());
  CHECK(filter.ok());
}

TEST_CASE("value = 32") {
  auto expr = equal(literal(32), field_ref("value"));
  auto bar = lance::arrow::ToArray({1, 2, 32, 0, 32}).ValueOrDie();
  auto struct_arr =
      ::arrow::StructArray::Make({bar}, {::arrow::field("value", ::arrow::int32())}).ValueOrDie();
  auto batch = ::arrow::RecordBatch::FromStructArray(struct_arr).ValueOrDie();

  auto filter = lance::io::Filter::Make(kSchema, expr).ValueOrDie();
  auto [indices, output] = filter->Exec(batch).ValueOrDie();
  CHECK(indices->Equals(lance::arrow::ToArray<uint64_t>({2, 4}).ValueOrDie()));

  bar = lance::arrow::ToArray({32, 32}).ValueOrDie();
  struct_arr =
      ::arrow::StructArray::Make({bar}, {::arrow::field("value", ::arrow::int32())}).ValueOrDie();
  batch = ::arrow::RecordBatch::FromStructArray(struct_arr).ValueOrDie();
  CHECK(output->Equals(*batch));
}

TEST_CASE("label = cat or label = dog") {
  auto expr =
      or_(equal(field_ref("label"), literal("cat")), equal(field_ref("label"), literal("dog")));
  auto labels =
      lance::arrow::ToArray({"person", "dog", "cat", "car", "cat", "food", "hotdog"}).ValueOrDie();
  auto struct_arr =
      ::arrow::StructArray::Make({labels}, {::arrow::field("label", ::arrow::utf8())}).ValueOrDie();
  auto batch = ::arrow::RecordBatch::FromStructArray(struct_arr).ValueOrDie();

  auto filter = lance::io::Filter::Make(kSchema, expr).ValueOrDie();
  auto [indices, output] = filter->Exec(batch).ValueOrDie();
  CHECK(indices->Equals(lance::arrow::ToArray<uint64_t>({1, 2, 4}).ValueOrDie()));

  labels = lance::arrow::ToArray({"dog", "cat", "cat"}).ValueOrDie();
  struct_arr =
      ::arrow::StructArray::Make({labels}, {::arrow::field("label", ::arrow::utf8())}).ValueOrDie();
  auto expected = ::arrow::RecordBatch::FromStructArray(struct_arr).ValueOrDie();
  CHECK(output->Equals(*expected));
}

File: java/core/src/test/java/com/lancedb/lance/FilterTest.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.lancedb.lance;

import java.io.IOException;
import java.nio.file.Path;

import com.lancedb.lance.ipc.LanceScanner;
import com.lancedb.lance.ipc.ScanOptions;
import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.memory.RootAllocator;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class FilterTest {
  @TempDir
  static Path tempDir;
  private static BufferAllocator allocator;
  private static Dataset dataset;

  @BeforeAll
  static void setup() throws IOException {
    String datasetPath = tempDir.resolve("filter_test_dataset").toString();
    allocator = new RootAllocator();
    TestUtils.SimpleTestDataset testDataset = new TestUtils.SimpleTestDataset(allocator, datasetPath);
    testDataset.createEmptyDataset().close();
    // write id with value from 0 to 39
    dataset = testDataset.write(1, 40);
  }

  @AfterAll
  static void tearDown() {
    // Cleanup resources used by the tests
    if (dataset != null) {
      dataset.close();
    }
    if (allocator != null) {
      allocator.close();
    }
  }

  @Test
  void testFilters() throws Exception {
    testFilter("id == 10", 1);
    testFilter("id == 10", 1);
    testFilter("id != 10", 39);
    testFilter("id > 10", 29);
    testFilter("id >= 10", 30);
    testFilter("id < 10", 10);
    testFilter("id <= 10", 11);
    testFilter("id >= 10 and id < 20", 10);
    testFilter("id < 10 or id > 30", 19);
    testFilter("id != 10 and id < 20", 19);
    testFilter("id < 5 or id > 35", 9);
    testFilter("(id >= 5 and id <= 15) or (id >= 25 and id <= 35)", 22);
    testFilter("id == 5 or (id >= 30 and id < 35)", 6);
    testFilter("id IS NOT NULL", 40);
    testFilter("id IS NULL", 0);
    testFilter("id IN (5, 15, 25, 35)", 4);

    testFilter("name LIKE 'Person%'", 40);
    testFilter("name LIKE 'Person 1%'", 11);
    testFilter("name LIKE '%0'", 4);
    testFilter("name LIKE '%son 1'", 1);
    testFilter("name LIKE '%son 1%'", 11);
    testFilter("name LIKE '%son%'", 40);
    testFilter("name == 'Person 1'", 1);
    testFilter("name IS NULL", 0);
    testFilter("name IS NOT NULL", 40);

    testFilter("name LIKE 'Person%' AND name LIKE '%0'", 4);
    testFilter("name LIKE 'Person%' AND name LIKE '%1%'", 13);
    testFilter("name LIKE '%son%' AND name LIKE '%0'", 4);
    testFilter("name LIKE '%son%' AND name LIKE '%1'", 4);
    testFilter("name LIKE 'Person 1%' AND name LIKE '%1'", 2);

    testFilter("(name IS NOT NULL) AND (name == 'Person 1')", 1);
    testFilter("(name IS NOT NULL) AND (name == 'Person')", 0);
    // Not supported, bug?, LanceError(IO): Schema error: No field named person. Valid fields are id, name.
    // testFilter("(name IS NOT NULL) AND (name == Person)", 0);

    // Not supported
    // testFilter("\"id\" == 10", 1);
    // testFilter("'id' == 10", 1);
  }

  private void testFilter(String filter, int expectedCount) throws Exception {
    try (LanceScanner scanner = dataset.newScan(new ScanOptions.Builder().filter(filter).build())) {
      assertEquals(expectedCount, scanner.countRows());
    }
  }
}
